rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp

```diff
@@ -0,0 +1,159 @@
+    _pool_seq = 0;
+    _exec_seq = 0;
+
+    for (size_t i = 0; i < _pool_size; ++i) {
+      _memory_map[_memory_pool[i]] = 0;
+    }
+
+    for (size_t i = 0; i < _max_executables; ++i) {
+      _executable_pool[i] = std::make_shared<executor::AnyExecutable>(executor::AnyExecutable());
+    }
+  }
+
+  void ** borrow_handles(HandleType type, size_t number_of_handles)
+  {
+    switch (type) {
+      case HandleType::subscriber_handle:
+        return _subscriber_pool;
+      case HandleType::service_handle:
+        return _service_pool;
+      case HandleType::client_handle:
+        return _client_pool;
+      case HandleType::guard_condition_handle:
+        return _guard_condition_pool;
+      default:
+        break;
+    }
+    throw std::runtime_error("Unrecognized enum, could not borrow handle memory.");
+  }
+
+  void return_handles(HandleType type, void ** handles)
+  {
+    switch (type) {
+      case HandleType::subscriber_handle:
+        memset(_subscriber_pool, 0, _max_subscribers);
+        break;
+      case HandleType::service_handle:
+        memset(_service_pool, 0, _max_services);
+        break;
+      case HandleType::client_handle:
+        memset(_client_pool, 0, _max_clients);
+        break;
+      case HandleType::guard_condition_handle:
+        memset(_guard_condition_pool, 0, _max_guard_conditions);
+        break;
+      default:
+        throw std::runtime_error("Unrecognized enum, could not return handle memory.");
+        break;
+    }
+  }
+
+  executor::AnyExecutableSharedPtr instantiate_next_executable()
+  {
+    if (_exec_seq >= _max_executables) {
+      // wrap around
+      _exec_seq = 0;
+    }
+    size_t prev_exec_seq = _exec_seq;
+    ++_exec_seq;
+
+    return _executable_pool[prev_exec_seq];
+  }
+
+  void * rcl_malloc(size_t size)
+  {
+    // Extremely naive static allocation strategy
+    // Keep track of block size at a given pointer
+    if (_pool_seq + size > _pool_size) {
+      // Start at 0
+      _pool_seq = 0;
+    }
+    void * ptr = _memory_pool[_pool_seq];
+    if (_memory_map.count(ptr) == 0) {
+      // We expect to have the state for all blocks pre-mapped into _memory_map
+      throw std::runtime_error("Unexpected pointer in rcl_malloc.");
+    }
+    _memory_map[ptr] = size;
+    size_t prev_pool_seq = _pool_seq;
+    _pool_seq += size;
+    return _memory_pool[prev_pool_seq];
+  }
+
+  void rcl_free(void * ptr)
+  {
+    if (_memory_map.count(ptr) == 0) {
+      // We expect to have the state for all blocks pre-mapped into _memory_map
+      throw std::runtime_error("Unexpected pointer in rcl_free.");
+    }
+
+    memset(ptr, 0, _memory_map[ptr]);
+  }
+
+protected:
+private:
+  static const size_t _pool_size = 1024;
+  static const size_t _max_subscribers = 10;
+  static const size_t _max_services = 5;
+  static const size_t _max_clients = 10;
+  static const size_t _max_guard_conditions = 50;
+  static const size_t _max_executables = 1;
```

Comment:
Yeah, I haven't hammered that out yet because I wanted to get this PR out
the door and hardcoded `static const` variables were the easiest way to
instantiate static arrays, but obviously it's not a good final solution.

This relates to my comment below about how `StaticMemoryAllocator` is
basically for demonstration purposes. For the application of this work,
which is real-time safe computing, it's okay to have dynamic memory
allocations in setup and teardown of the program, which is the easiest way
to make these parameters configurable.

But, if we want to build a purely static memory allocator for the purposes
of demonstration, here are two ideas:

Pass the parameters to a code generation module (parameters are determined
at compile time).

Get parameters at runtime (from command line, parameter server, whatever).
Reserve a maximum memory pool with a hardcoded value and divide up the pool
based on parameters and size of objects. If maximum pool size is exceeded,
throw an error.


On Mon, Jul 13, 2015 at 3:10 PM, William Woodall <notifications@github.com>
wrote:

> In rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp
> <https://github.com/ros2/rclcpp/pull/56#discussion_r34518109>:
>
> > +    if (_memory_map.count(ptr) == 0) {
> > +      // We expect to have the state for all blocks pre-mapped into _memory_map
> > +      throw std::runtime_error("Unexpected pointer in rcl_free.");
> > +    }
> > +
> > +    memset(ptr, 0, _memory_map[ptr]);
> > +  }
> > +
> > +protected:
> > +private:
> > +  static const size_t _pool_size = 1024;
> > +  static const size_t _max_subscribers = 10;
> > +  static const size_t _max_services = 5;
> > +  static const size_t _max_clients = 10;
> > +  static const size_t _max_guard_conditions = 50;
> > +  static const size_t _max_executables = 1;
>
> At some point we have to figure out how to make this stuff configurable.
> Any ideas on that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ros2/rclcpp/pull/56/files#r34518109>.
>